File: server_old/conversations/Writing.js
import OpenAI from 'openai';
import 'dotenv/config';
import { SystemMessage } from '../prompts/prompt_Writing.js';
import { generateSystemMessage } from '../prompts/prompt_Summarize.js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

export class Author {
    constructor(apiKey) {
        this.OpenAI = new OpenAI({ apiKey });
        this.systemMessage = SystemMessage;
        this.chapters = [];
    }

    async sendChatCompletion(message) {
        return await this.OpenAI.chat.completions.create({
            model: "gpt-4-turbo-preview",
            messages: [
                { "role": "system", "content": this.systemMessage },
                { "role": "user", "content": message }
            ],
            temperature: 1,
            max_tokens: 4096,
            top_p: 1,
            frequency_penalty: 0,
            presence_penalty: 0,
            response_format: { "type": "json_object" }
        });
    }
}

class Summarizer {
    constructor(apiKey) {
        this.OpenAI = new OpenAI({ apiKey });
    }

    async sendChatCompletion(chapter) {
        let systemMessage = generateSystemMessage(chapter);
        return await this.OpenAI.chat.completions.create({
            model: "gpt-4-turbo-preview",
            messages: [
                { "role": "system", "content": systemMessage },
            ],
            temperature: 1,
            max_tokens: 4096,
            top_p: 1,
            frequency_penalty: 0,
            presence_penalty: 0,
            response_format: { "type": "text" }
        });
    }
}

async function run() {
    let outline = JSON.parse(fs.readFileSync('./BookOutlines/BookOutline.json', 'utf8'));
    let plan = JSON.parse(fs.readFileSync('./BookPlans/BookPlan.json', 'utf8'));
    let priorSummaries = [];
    let author = new Author(process.env.OPENAI_API_KEY);
    let summarizer = new Summarizer(process.env.OPENAI_API_KEY);
    for (let i = 0; i < outline.length; i++) {
        console.log(`Now writing: Chapter ${i + 1}`)
        let message = JSON.stringify({"bookPlan" : plan["bookPlan"], "chatperOutline": outline[i], "priorSummaries": JSON.stringify(priorSummaries)});
        console.log(`Chapter ${i + 1} message: ${message}`)
        let response = await author.sendChatCompletion(message);
        author.chapters.push(JSON.parse(response.choices[0].message.content));
        let summary = await summarizer.sendChatCompletion(response.choices[0].message.content);
        console.log(`Chapter ${i + 1} Summary: ${summary.choices[0].message.content}`);
        let summarykey = `Chapter ${i + 1}`;
        priorSummaries.push({[summarykey] : summary.choices[0].message.content});
    //console.log(response.choices[0].message.content);
    //console.log(summary.choices[0].message.content);
    }
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = dirname(__filename);

    let filePath = './Book/Book.json';
    fs.writeFileSync(filePath, JSON.stringify(author.chapters));
}

run();

File: server/__openai/writing.py

from typing import List, Dict
import json
from sqlmodel import Session
from config import settings

from models.openai import Message, ChatCompletion
from models.database import BookPlan, ChapterOutlines, Chapter

from __openai.utils import OpenAIBase
from __openai.prompts.prompt_writing import systemMessage as AuthorPrompt
from __openai.prompts.prompt_summarizing import systemMessage as SummarizerPrompt

from database.crud import (
    get_bookplan,
    get_chapter_outlines
    )
from database.database import engine

class Summarizer(OpenAIBase):
    def __init__(self, api_key: str):
        super().__init__(api_key)

    def send_chat_completion(self, chapter: str) -> ChatCompletion:
        system_message = SummarizerPrompt.format(chapter=chapter)
        return super().send_chat_completion(
            messages=[
            Message(role="system", content=system_message),
            ],
            model="gpt-3.5-turbo-0125",
            temperature=1,
            max_tokens=4096,
            top_p=1,
            frequency_penalty=0,
            presence_penalty=0,
            response_format={"type": "text"}
        )

class Author(OpenAIBase):
    def __init__(self, api_key: str, bookplan_id: str):
        super().__init__(api_key)
        self.system_message: str = AuthorPrompt
        self.chapters: List[Chapter] = []
        self.bookplan_id = bookplan_id

    def send_chat_completion(self, message: Message) -> ChatCompletion:
        return super().send_chat_completion(
            messages=[
                Message(role="system", content=self.system_message),
                Message(role="user", content=message.content)
            ],
            temperature=1,
            max_tokens=4096,
            top_p=1,
            frequency_penalty=0,
            presence_penalty=0,
            response_format={"type": "json_object"}
        )

    def write(self) -> None:
        with Session(engine) as db:
            plan: BookPlan = get_bookplan(db=db, bookplan_id=self.bookplan_id)
            outlines: List[ChapterOutlines] = get_chapter_outlines(db=db, bookplan_id=self.bookplan_id)
        
        prior_summaries = []
        
        for i, chapter_outline in enumerate(outlines):
            print(f"Now writing: Chapter {i + 1}")
            message_content = json.dumps({
                "bookPlan": plan.Book_Plan,
                "chapterOutline": chapter_outline.Chapter_Outline_JSON,
                "priorSummaries": prior_summaries
            })
            print(f"Chapter {i + 1} message: {message_content}")
            
            message = Message(role="user", content=message_content)

            response = self.send_chat_completion(message)
            chapter = Chapter(ChapterOutline_ID=chapter_outline.ChapterOutline_ID, 
                              BookPlan_ID=chapter_outline.BookPlan_ID,
                              Chapter_Num=chapter_outline.Chapter_Num,
                              Chapter=json.loads(response.choices.message.content)['Chapter'])
            self.chapters.append(chapter)
   
            summary = Summarizer(settings.openai_api_key).send_chat_completion(response.choices.message.content)
            summary_text = summary.choices.message.content
            print(f"Chapter {i + 1} Summary: {summary_text}")
            
            summary_key = f"Chapter {i + 1}"
            prior_summaries.append({summary_key: summary_text})
        
